### ISSUE TITLE
TLS/PSK backend possible?

### ISSUE THREAD

**Issue by nutinshell** [nghttp2/nghttp2#772]
In some scenario, we have to use remote backend, for security and performance, could we achieve like this:

user -- (tls) -- nghttpx -- (maybe tls-psk even with tfo?) -- remote-nghttpx -- plaintext-backend 

Thanks for any advice.

**Comment by tatsuhiro-t** [comment]
Backend connections can be encrypted using "tls" parameter in backend option.  nghttpx does not support PSK yet.  I'm not sure it is entirely possible with OpenSSL cause its use seems to be rare at the moment.  See also #612

**Comment by nutinshell** [comment]
thanks for the info. stunnel has this feature, https://www.stunnel.org/auth.html, it should be very easy to implement with openssl I think.

**Comment by tatsuhiro-t** [comment]
Thank you.  Will check stunnel.

**Comment by tatsuhiro-t** [comment]
#774 implements PSK in nghttpx.

**Comment by nutinshell** [comment]
just had a quick test, extra "WARN (shrpx_config.cc:2455) frontend-no-tls: deprecated.  Use no-tls keyword in frontend" when use "--psk-secrets" and still require key and cert.

**Comment by tatsuhiro-t** [comment]
frontend-no-tls has been deprecated while ago.  Follow the warning message to fix it.

Currently certificate and private key are still required even if enabled cipher suites are PSK only.  Specify dummy self signed cert if PSK is solely used.

**Comment by nutinshell** [comment]
seems not so normal, never called frontend-no-tls, and -b ignored.
```
nghttpx -n 2 -f'0.0.0.0,10001' -b'127.0.0.1,3228' --tls-proto-list=TLSv1.2 --psk-secrets=/etc/ngpsk --ciphers=PSK-AES128-GCM-SHA256 s.key s.pem 
2017-01-07T17:34:45.944Z 41626 41626 e60b82b6 WARN (shrpx_config.cc:2455) frontend-no-tls: deprecated.  Use no-tls keyword in frontend
2017-01-07T17:34:45.944Z 41626 41626 e60b82b6 WARN (shrpx.cc:2645) --fetch-ocsp-response-file: /build/ngpsk/share/nghttp2/fetch-ocsp-response not found.  OCSP stapling has been disabled.
2017-01-07T17:34:45.944Z 41626 41626 e60b82b6 NOTICE (shrpx.cc:819) Listening on 127.0.0.1:1000, tls
2017-01-07T17:34:45.944Z 41626 41626 e60b82b6 NOTICE (shrpx.cc:819) Listening on 0.0.0.0:60001, tls
2017-01-07T17:34:45.944Z 41626 41626 e60b82b6 NOTICE (shrpx.cc:1218) Worker process [41627] spawned
```

**Comment by tatsuhiro-t** [comment]
Make sure that you don't have nghttpx configuration file /etc/nghttpx/nghttpx.conf.

**Comment by nutinshell** [comment]
simply fixed that :)

setup like below, curl https://y.y.y.y:1000 return 502 error, any thing wrong I made?

client: `nghttpx -s -f'y.y.y.y,1000' -b'x.x.x.x,2000;tls' --tls-proto-list=TLSv1.2 --client-psk-secrets=/etc/ngpsk  s.key s.pem`

server: `nghttpx -n 2 -f'0.0.0.0,2000' -b'127.0.0.1,3228' --tls-proto-list=TLSv1.2 --psk-secrets=/etc/ngpsk --ciphers=PSK-AES128-GCM-SHA256 s.key s.pem`

**Comment by tatsuhiro-t** [comment]
nghttpx client also should include PSK-AES128-GCM-SHA256 using --ciphers.  Note that we will probably add new option --client-ciphers, which only affects backend cipher suites, and --ciphers would change only frontend ciphers.

**Comment by nutinshell** [comment]
--client-ciphers makes sense, and we could terminate SSL backend into plaintext frontend with no-tls.

**Comment by tatsuhiro-t** [comment]
I will add --client-ciphers after merging this stuff, since it could make a conflict.
Other than the that, did you notice anything?  Does it work in a good shape?

**Comment by nutinshell** [comment]
runs good, but still can't make it work in real world since if config with PSK-AES128-GCM-SHA256, no one browser accept it, we should have --client-ciphers first :)

**Comment by tatsuhiro-t** [comment]
Well, you can just append PSK-AES128-GCM-SHA256 to the cipher suite list, leaving default cipher suites for browsers.

**Comment by nutinshell** [comment]
like `--ciphers=ECDHE-RSA-AES128-GCM-SHA256:PSK-AES128-GCM-SHA256` ? 

same 502 error.

**Comment by tatsuhiro-t** [comment]
Yes, for client nghttpx.
Command-line?

**Comment by nutinshell** [comment]
it's `nghttpx -f'0.0.0.0,1000' -b'x.x.x.x,2000;tls' --tls-proto-list=TLSv1.2 --client-psk-secrets=/etc/ngpsk  ../s.key ../s.pem -s --ciphers=ECDHE-RSA-AES128-GCM-SHA256:PSK-AES128-GCM-SHA256`

**Comment by tatsuhiro-t** [comment]
Thank you.  PSK-AES128-GCM-SHA256 is black listed by HTTP/2, no-http2-cipher-black-list option must be used with PSK.  Due to bug, the option does not work with HTTP/2 backend connection.  It was fixed by the commit 0abc220013d084283ed3f65de24f8b36673cb522
